File: hangman/hangman.py
import pygame
import random
import sys

# Monokai Theme Colors
COLOR_BG = (39, 40, 34)       # #272822
COLOR_FG = (248, 248, 242)    # #F8F8F2
COLOR_YELLOW = (230, 219, 116) # #E6DB74
COLOR_MAGENTA = (249, 38, 114) # #F92672
COLOR_CYAN = (102, 217, 239)    # #66D9EF
COLOR_GREEN = (166, 226, 46)    # #A6E22E
COLOR_ORANGE = (253, 151, 31)   # #FD971F

SCREEN_WIDTH = 800
SCREEN_HEIGHT = 600
FPS = 60

def get_word():
    words = ["PYTHON", "CPLUSPLUS", "SFML", "PROGRAMMING", "DEVELOPER", "COMPUTER", "NETWORK", "ALGORITHM", "SENSORS", "SYNTHWAVE"]
    return random.choice(words).upper()

class HangmanGame:
    def __init__(self):
        pygame.init()
        self.screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
        pygame.display.set_caption("MONOKAI HANGMAN")
        self.clock = pygame.time.Clock()
        
        try:
            self.font = pygame.font.Font("/usr/share/fonts/libertinus/LibertinusSans-Bold.ttf", 32)
            self.small_font = pygame.font.Font("/usr/share/fonts/libertinus/LibertinusSans-Bold.ttf", 24)
        except:
            self.font = pygame.font.SysFont("Arial", 32, bold=True)
            self.small_font = pygame.font.SysFont("Arial", 24, bold=True)

        self.reset()

    def reset(self):
        self.word = get_word()
        self.guessed_letters = set()
        self.tries = 0
        self.max_tries = 6
        self.game_state = "PLAYING" # PLAYING, WON, LOST

    def draw_hangman(self):
        # Gallow
        pygame.draw.line(self.screen, COLOR_FG, (100, 500), (300, 500), 5) # Base
        pygame.draw.line(self.screen, COLOR_FG, (200, 500), (200, 100), 5) # Post
        pygame.draw.line(self.screen, COLOR_FG, (200, 100), (300, 100), 5) # Beam
        pygame.draw.line(self.screen, COLOR_FG, (300, 100), (300, 150), 5) # Rope

        if self.tries >= 1: # Head
            pygame.draw.circle(self.screen, COLOR_MAGENTA, (300, 170), 20, 5)
        if self.tries >= 2: # Body
            pygame.draw.line(self.screen, COLOR_MAGENTA, (300, 190), (300, 300), 5)
        if self.tries >= 3: # Left Arm
            pygame.draw.line(self.screen, COLOR_MAGENTA, (300, 210), (260, 250), 5)
        if self.tries >= 4: # Right Arm
            pygame.draw.line(self.screen, COLOR_MAGENTA, (300, 210), (340, 250), 5)
        if self.tries >= 5: # Left Leg
            pygame.draw.line(self.screen, COLOR_MAGENTA, (300, 300), (260, 350), 5)
        if self.tries >= 6: # Right Leg
            pygame.draw.line(self.screen, COLOR_MAGENTA, (300, 300), (340, 350), 5)

    def draw_word(self):
        display_text = ""
        for char in self.word:
            if char in self.guessed_letters:
                display_text += char + " "
            else:
                display_text += "_ "
        
        text_surf = self.font.render(display_text, True, COLOR_CYAN)
        text_rect = text_surf.get_rect(center=(SCREEN_WIDTH // 2, 400))
        self.screen.blit(text_surf, text_rect)

    def draw_keyboard(self):
        alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        start_x = 100
        start_y = 460
        btn_w = 40
        btn_h = 40
        gap = 10

        for i, char in enumerate(alphabet):
            row = i // 13
            col = i % 13
            rect = pygame.Rect(start_x + col * (btn_w + gap), start_y + row * (btn_h + gap), btn_w, btn_h)
            
            color = COLOR_BG
            if char in self.guessed_letters:
                color = (60, 60, 50) # Darker
            
            pygame.draw.rect(self.screen, color, rect, border_radius=5)
            pygame.draw.rect(self.screen, COLOR_FG, rect, 2, border_radius=5)
            
            txt = self.small_font.render(char, True, COLOR_YELLOW if char not in self.guessed_letters else COLOR_FG)
            txt_rect = txt.get_rect(center=rect.center)
            self.screen.blit(txt, txt_rect)
            
            # Return rects for collision detection in the run loop
            # (Actually easier to just calculate based on mouse pos)

    def draw_overlay(self):
        overlay = pygame.Surface((SCREEN_WIDTH, SCREEN_HEIGHT), pygame.SRCALPHA)
        overlay.fill((0, 0, 0, 180))
        self.screen.blit(overlay, (0, 0))
        
        msg = "YOU WON!" if self.game_state == "WON" else "GAME OVER"
        color = COLOR_GREEN if self.game_state == "WON" else COLOR_MAGENTA
        
        res_text = self.font.render(msg, True, color)
        res_rect = res_text.get_rect(center=(SCREEN_WIDTH // 2, SCREEN_HEIGHT // 2 - 20))
        self.screen.blit(res_text, res_rect)
        
        word_text = self.font.render(f"Word was: {self.word}", True, COLOR_FG)
        word_rect = word_text.get_rect(center=(SCREEN_WIDTH // 2, SCREEN_HEIGHT // 2 + 20))
        self.screen.blit(word_text, word_rect)
        
        restart_text = self.small_font.render("Press 'R' to Restart", True, COLOR_CYAN)
        restart_rect = restart_text.get_rect(center=(SCREEN_WIDTH // 2, SCREEN_HEIGHT // 2 + 60))
        self.screen.blit(restart_text, restart_rect)

    def run(self):
        running = True
        while running:
            self.screen.fill(COLOR_BG)
            
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    running = False
                
                if event.type == pygame.KEYDOWN:
                    if event.key == pygame.K_r:
                        self.reset()
                
                if event.type == pygame.MOUSEBUTTONDOWN and self.game_state == "PLAYING":
                    mx, my = pygame.mouse.get_pos()
                    # Check keyboard clicks
                    start_x = 100
                    start_y = 460
                    btn_w = 40
                    btn_h = 40
                    gap = 10
                    
                    alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
                    for i, char in enumerate(alphabet):
                        row = i // 13
                        col = i % 13
                        rect = pygame.Rect(start_x + col * (btn_w + gap), start_y + row * (btn_h + gap), btn_w, btn_h)
                        if rect.collidepoint(mx, my) and char not in self.guessed_letters:
                            self.guessed_letters.add(char)
                            if char not in self.word:
                                self.tries += 1
                            if all(c in self.guessed_letters for c in self.word):
                                self.game_state = "WON"
                            elif self.tries >= self.max_tries:
                                self.game_state = "LOST"

            self.draw_hangman()
            self.draw_word()
            self.draw_keyboard()
            
            if self.game_state != "PLAYING":
                self.draw_overlay()
            
            pygame.display.flip()
            self.clock.tick(FPS)
            
        pygame.quit()

if __name__ == "__main__":
    HangmanGame().run()

File: hangman/hangman.cpp
#include <SFML/Graphics.hpp>
#include <iostream>
#include <vector>
#include <string>
#include <random>
#include <set>
#include <algorithm>
#include <cmath>
#include <optional>

// Monokai Theme Colors
const sf::Color COLOR_BG(39, 40, 34);
const sf::Color COLOR_FG(248, 248, 242);
const sf::Color COLOR_YELLOW(230, 219, 116);
const sf::Color COLOR_MAGENTA(249, 38, 114);
const sf::Color COLOR_CYAN(102, 217, 239);
const sf::Color COLOR_GREEN(166, 226, 46);
const sf::Color COLOR_ORANGE(253, 151, 31);

const int SCREEN_WIDTH = 800;
const int SCREEN_HEIGHT = 600;

std::string getWord() {
    std::vector<std::string> words = {"PYTHON", "CPLUSPLUS", "SFML", "PROGRAMMING", "DEVELOPER", "COMPUTER", "NETWORK", "ALGORITHM", "SENSORS", "SYNTHWAVE"};
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, (int)words.size() - 1);
    return words[dis(gen)];
}

class HangmanGame {
public:
    HangmanGame() {
        window.create(sf::VideoMode({SCREEN_WIDTH, SCREEN_HEIGHT}), "MONOKAI HANGMAN");
        
        if (!font.openFromFile("/usr/share/fonts/libertinus/LibertinusSans-Bold.ttf")) {
            std::cerr << "Could not load font, please ensure /usr/share/fonts/libertinus/LibertinusSans-Bold.ttf exists." << std::endl;
        }
        
        reset();
    }

    void reset() {
        word = getWord();
        guessedLetters.clear();
        tries = 0;
        maxTries = 6;
        gameState = "PLAYING";
    }

    void drawLine(float x1, float y1, float x2, float y2, float thickness, sf::Color color) {
        sf::RectangleShape line;
        float dx = x2 - x1;
        float dy = y2 - y1;
        float length = std::sqrt(dx * dx + dy * dy);
        line.setSize({length, thickness});
        line.setRotation(sf::degrees(std::atan2(dy, dx) * 180.0f / 3.14159f));
        line.setPosition({x1, y1});
        line.setFillColor(color);
        window.draw(line);
    }

    void drawHangman() {
        drawLine(100, 500, 300, 500, 5, COLOR_FG);
        drawLine(200, 500, 200, 100, 5, COLOR_FG);
        drawLine(200, 100, 300, 100, 5, COLOR_FG);
        drawLine(300, 100, 300, 150, 5, COLOR_FG);

        if (tries >= 1) {
            sf::CircleShape head(20);
            head.setOutlineThickness(5);
            head.setOutlineColor(COLOR_MAGENTA);
            head.setFillColor(sf::Color::Transparent);
            head.setPosition({300.0f - 20.0f, 170.0f - 20.0f});
            window.draw(head);
        }
        if (tries >= 2) drawLine(300, 190, 300, 300, 5, COLOR_MAGENTA);
        if (tries >= 3) drawLine(300, 210, 260, 250, 5, COLOR_MAGENTA);
        if (tries >= 4) drawLine(300, 210, 340, 250, 5, COLOR_MAGENTA);
        if (tries >= 5) drawLine(300, 300, 260, 350, 5, COLOR_MAGENTA);
        if (tries >= 6) drawLine(300, 300, 340, 350, 5, COLOR_MAGENTA);
    }

    void drawWord() {
        std::string displayText = "";
        for (char c : word) {
            if (guessedLetters.count(c)) {
                displayText += c;
                displayText += " ";
            } else {
                displayText += "_ ";
            }
        }
        sf::Text text(font, displayText, 32);
        text.setFillColor(COLOR_CYAN);
        sf::FloatRect textRect = text.getLocalBounds();
        text.setOrigin({textRect.position.x + textRect.size.x / 2.0f, textRect.position.y + textRect.size.y / 2.0f});
        text.setPosition({SCREEN_WIDTH / 2.0f, 400.0f});
        window.draw(text);
    }

    void drawKeyboard() {
        std::string alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        float startX = 100, startY = 460;
        float btnW = 40, btnH = 40, gap = 10;

        for (int i = 0; i < (int)alphabet.length(); ++i) {
            char c = alphabet[i];
            int row = i / 13;
            int col = i % 13;
            float x = startX + col * (btnW + gap);
            float y = startY + row * (btnH + gap);

            sf::RectangleShape rect({btnW, btnH});
            rect.setPosition({x, y});
            rect.setFillColor(guessedLetters.count(c) ? sf::Color(60, 60, 50) : COLOR_BG);
            rect.setOutlineThickness(2);
            rect.setOutlineColor(COLOR_FG);
            window.draw(rect);

            sf::Text text(font, std::string(1, c), 24);
            text.setFillColor(guessedLetters.count(c) ? COLOR_FG : COLOR_YELLOW);
            sf::FloatRect textRect = text.getLocalBounds();
            text.setOrigin({textRect.position.x + textRect.size.x / 2.0f, textRect.position.y + textRect.size.y / 2.0f});
            text.setPosition({x + btnW / 2.0f, y + btnH / 2.0f});
            window.draw(text);
        }
    }

    void drawOverlay() {
        sf::RectangleShape overlay({(float)SCREEN_WIDTH, (float)SCREEN_HEIGHT});
        overlay.setFillColor(sf::Color(0, 0, 0, 180));
        window.draw(overlay);

        std::string msg = (gameState == "WON") ? "YOU WON!" : "GAME OVER";
        sf::Color color = (gameState == "WON") ? COLOR_GREEN : COLOR_MAGENTA;

        sf::Text resText(font, msg, 32);
        resText.setFillColor(color);
        sf::FloatRect resRect = resText.getLocalBounds();
        resText.setOrigin({resRect.position.x + resRect.size.x / 2.0f, resRect.position.y + resRect.size.y / 2.0f});
        resText.setPosition({SCREEN_WIDTH / 2.0f, SCREEN_HEIGHT / 2.0f - 20.0f});
        window.draw(resText);

        sf::Text wordText(font, "Word was: " + word, 32);
        wordText.setFillColor(COLOR_FG);
        sf::FloatRect wRect = wordText.getLocalBounds();
        wordText.setOrigin({wRect.position.x + wRect.size.x / 2.0f, wRect.position.y + wRect.size.y / 2.0f});
        wordText.setPosition({SCREEN_WIDTH / 2.0f, SCREEN_HEIGHT / 2.0f + 20.0f});
        window.draw(wordText);

        sf::Text restartText(font, "Press 'R' to Restart", 24);
        restartText.setFillColor(COLOR_CYAN);
        sf::FloatRect rRect = restartText.getLocalBounds();
        restartText.setOrigin({rRect.position.x + rRect.size.x / 2.0f, rRect.position.y + rRect.size.y / 2.0f});
        restartText.setPosition({SCREEN_WIDTH / 2.0f, SCREEN_HEIGHT / 2.0f + 60.0f});
        window.draw(restartText);
    }

    void run() {
        while (window.isOpen()) {
            while (const std::optional event = window.pollEvent()) {
                if (event->is<sf::Event::Closed>())
                    window.close();
                
                if (const auto* keyPressed = event->getIf<sf::Event::KeyPressed>()) {
                    if (keyPressed->code == sf::Keyboard::Key::R) {
                        reset();
                    }
                }

                if (const auto* mousePressed = event->getIf<sf::Event::MouseButtonPressed>()) {
                    if (mousePressed->button == sf::Mouse::Button::Left && gameState == "PLAYING") {
                        sf::Vector2i mousePos = sf::Mouse::getPosition(window);
                        float mx = (float)mousePos.x;
                        float my = (float)mousePos.y;
                        
                        std::string alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
                        float startX = 100, startY = 460, btnW = 40, btnH = 40, gap = 10;
                        
                        for (int i = 0; i < (int)alphabet.length(); ++i) {
                            char c = alphabet[i];
                            int row = i / 13;
                            int col = i % 13;
                            float x = startX + col * (btnW + gap);
                            float y = startY + row * (btnH + gap);
                            
                            if (mx >= x && mx <= x + btnW && my >= y && my <= y + btnH && !guessedLetters.count(c)) {
                                guessedLetters.insert(c);
                                if (word.find(c) == std::string::npos) {
                                    tries++;
                                }
                                
                                bool won = true;
                                for (char wc : word) {
                                    if (!guessedLetters.count(wc)) {
                                        won = false;
                                        break;
                                    }
                                }
                                if (won) gameState = "WON";
                                else if (tries >= maxTries) gameState = "LOST";
                            }
                        }
                    }
                }
            }

            window.clear(COLOR_BG);
            drawHangman();
            drawWord();
            drawKeyboard();
            if (gameState != "PLAYING") drawOverlay();
            window.display();
        }
    }

private:
    sf::RenderWindow window;
    sf::Font font;
    std::string word;
    std::set<char> guessedLetters;
    int tries;
    int maxTries;
    std::string gameState;
};

int main() {
    HangmanGame game;
    game.run();
    return 0;
}
